xformers/triton/dropout.py

```diff
@@ -19,49 +19,58 @@
     get_triton_activation_kernel,
 )
 from xformers.triton.k_dropout import k_dropout_bw, k_dropout_fw
-from xformers.triton.sum_strided import sum_2d_dim_0
+
+GROUP_M = 32  # 32
+BLOCK_M = GROUP_M // 4
+BLOCK_N = 128  # 128
 
 
 # Helper to handle the SPMD launch grid and error cases
 class _dropout(torch.autograd.Function):
     @staticmethod
     @custom_fwd(cast_inputs=torch.float16)
-    def forward(ctx, x, p, bias, activation, activation_grad):
+    def forward(ctx, x, p, bias, activation, activation_grad, trainable_bias):
         # Soft-flatten an hypothetical 3rd dimension
         x_ = x.reshape(-1, x.shape[-1]).contiguous()
         y = torch.empty_like(x_)
-        _, N = x_.shape
-
-        assert bias is None or bias.dtype == x.dtype, bias
+        M, N = x_.shape
 
-        # Generate one seed per sample
-        # seed max is int32 max for positive numbers: 2**16
-        seeds = torch.randint(65536, (x_.shape[0],), device=x.device).to(torch.int32)
+        assert bias is None or (bias.dtype == x.dtype and bias.shape[0] == N)
 
-        # SPMD launch grid
         def grid(meta):
             return (
-                x_.shape[0],
-                triton.cdiv(x_.shape[1], meta["BLOCK_SIZE"]),
+                triton.cdiv(M, meta["BLOCK_M"] * 4),
```

Comment:
How come the number 4 here?

Reply:
Great question, I'll add a note ! It's because optimally the rand generator from Triton returns 4 blocks (you can see that in the kernel, we manually go through them since indexing is not possible), so the real footprint here is 4 times this base setting